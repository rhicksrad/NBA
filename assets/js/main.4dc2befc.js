var B=0;function T(n,e,t){var L,M;if(n.innerHTML="",!e||e.length===0){let p=document.createElement("p");p.textContent="Trend data will appear once the next game is complete.",n.appendChild(p);return}let r=(L=t.width)!=null?L:320,s=(M=t.height)!=null?M:120,a=12,o=e.map(p=>p.value),l=Math.min(...o),u=Math.max(...o)-l||1,g=p=>a+p/Math.max(e.length-1,1)*(r-a*2),x=p=>a+(1-(p-l)/u)*(s-a*2),h=[],v=[];e.forEach((p,y)=>{let E=g(y),U=x(p.value);v.push(`${y===0?"M":"L"}${E},${U}`),y===0?(h.push(`M${E},${s-a}`),h.push(`L${E},${U}`)):h.push(`L${E},${U}`),y===e.length-1&&(h.push(`L${E},${s-a}`),h.push("Z"))});let c=document.createElementNS("http://www.w3.org/2000/svg","svg"),b=++B;c.setAttribute("viewBox",`0 0 ${r} ${s}`),c.setAttribute("role","img"),c.setAttribute("aria-labelledby",`sparkline-title-${b} sparkline-desc-${b}`);let $=document.createElementNS(c.namespaceURI,"title");$.setAttribute("id",`sparkline-title-${b}`),$.textContent=t.title;let f=document.createElementNS(c.namespaceURI,"desc");f.setAttribute("id",`sparkline-desc-${b}`),f.textContent=t.description;let m=document.createElementNS(c.namespaceURI,"path");m.setAttribute("d",h.join(" ")),m.setAttribute("class","sparkline-fill");let d=document.createElementNS(c.namespaceURI,"path");d.setAttribute("d",v.join(" ")),d.setAttribute("class","sparkline-path");let S=e[e.length-1],C=document.createElementNS(c.namespaceURI,"circle");C.setAttribute("cx",String(g(e.length-1))),C.setAttribute("cy",String(x(S.value))),C.setAttribute("r","4"),C.setAttribute("class","sparkline-point"),c.append($,f,m,d,C),n.appendChild(c)}var H=0;function I(n,e,t){var h,v;if(n.innerHTML="",!e||e.length===0){let c=document.createElement("p");c.textContent="Comparison data not yet available.",n.appendChild(c);return}let r=(h=t.width)!=null?h:360,s=(v=t.height)!=null?v:160,a={top:16,right:24,bottom:24,left:140},o=Math.max(...e.map(c=>c.value),1),l=(s-a.top-a.bottom)/e.length-12,i=document.createElementNS("http://www.w3.org/2000/svg","svg"),u=++H;i.setAttribute("viewBox",`0 0 ${r} ${s}`),i.setAttribute("role","img"),i.setAttribute("aria-labelledby",`bar-title-${u} bar-desc-${u}`);let g=document.createElementNS(i.namespaceURI,"title");g.setAttribute("id",`bar-title-${u}`),g.textContent=t.title;let x=document.createElementNS(i.namespaceURI,"desc");x.setAttribute("id",`bar-desc-${u}`),x.textContent=t.description,i.append(g,x),e.forEach((c,b)=>{let $=document.createElementNS(i.namespaceURI,"g"),f=a.top+b*(l+12),m=document.createElementNS(i.namespaceURI,"text");m.setAttribute("x",String(a.left-12)),m.setAttribute("y",String(f+l/2+6)),m.setAttribute("class","bar-label"),m.setAttribute("text-anchor","end"),m.textContent=c.label;let d=document.createElementNS(i.namespaceURI,"rect");d.setAttribute("x",String(a.left)),d.setAttribute("y",String(f)),d.setAttribute("width",String((r-a.left-a.right)*c.value/o)),d.setAttribute("height",String(Math.max(l,8))),d.setAttribute("rx","6"),d.setAttribute("class",b===0?"bar-rect":"bar-rect-secondary");let S=document.createElementNS(i.namespaceURI,"text");S.setAttribute("x",String(a.left+(r-a.left-a.right)*c.value/o+8)),S.setAttribute("y",String(f+l/2+6)),S.setAttribute("class","bar-value"),S.textContent=`${c.value}`,$.append(d,m,S),i.appendChild($)}),n.appendChild(i)}function N(n){let e=document.createElement("a");e.className="source-badge",e.href=n.url,e.target="_blank",e.rel="noopener",e.setAttribute("aria-label",`${n.name} \u2014 ${n.updateCadence}`),e.textContent=`Source: ${n.name}`;let t=document.createElement("span");return t.textContent=n.updateCadence,e.appendChild(t),e}function R(n){if(!n)return"Last updated: Not available";let e=new Date(n);if(Number.isNaN(e.getTime()))return"Last updated: Invalid date";let t=i=>i.toString().padStart(2,"0"),r=e.getUTCFullYear(),s=t(e.getUTCMonth()+1),a=t(e.getUTCDate()),o=t(e.getUTCHours()),l=t(e.getUTCMinutes());return`Last updated: ${r}-${s}-${a} ${o}:${l} UTC`}async function w(n){try{let e=await fetch(n,{cache:"no-cache"});if(!e.ok)throw new Error(`Failed to load ${n}: ${e.status}`);return await e.json()}catch(e){return console.error(e),null}}function k(n,e){let t=document.querySelector(`[data-source="${n}"]`);if(t){if(t.innerHTML="",!e){t.textContent="Source unavailable";return}t.appendChild(N(e))}}function A(n,e){let t=document.querySelector(`[data-last-updated="${n}"]`);t&&(t.textContent=R(e!=null?e:void 0))}async function P(){let n=document.querySelector('[data-module="power-status"]');if(!n)return;let e=await w("data/power-status.json");if(!e){n.textContent="Status indicators are temporarily unavailable.";return}A("power-status",e.lastUpdated),k("power-status",e.source),n.innerHTML="",e.items.forEach(t=>{var g;let r=document.createElement("article");r.className="status-card",r.tabIndex=0,r.setAttribute("aria-label",`${t.label} is ${t.status} at ${t.value} ${t.unit}`);let s=document.createElement("h3");s.textContent=t.label;let a=document.createElement("span");a.className=`status-tag status-${t.status.toLowerCase()}`,a.textContent=t.status;let o=document.createElement("div");o.className="value",o.textContent=`${t.value.toFixed(1)} ${t.unit}`;let l=document.createElement("p"),i=t.change>0?"+":"";l.textContent=`${i}${t.change.toFixed(1)} vs last check`;let u=document.createElement("p");u.textContent=(g=t.description)!=null?g:"",r.append(s,a,o,l,u),n.appendChild(r)})}async function D(){var r,s;let n=document.querySelector('[data-chart="sparkline"]');if(!n)return;let e=document.getElementById("sparkline-desc"),t=await w("data/team-sparkline.json");if(!t){n.textContent="Unable to load recent milestones.",e&&(e.textContent="");return}A("sparkline",t.lastUpdated),k("sparkline",t.source),T(n,t.values,{title:`${t.team} ${t.metric}`,description:(r=t.summary)!=null?r:`${t.team} ${t.metric}`}),e&&(e.textContent=(s=t.summary)!=null?s:"")}async function j(){var r,s;let n=document.querySelector('[data-chart="bar-chart"]');if(!n)return;let e=document.getElementById("bar-chart-desc"),t=await w("data/conference-bar.json");if(!t){n.textContent="Unable to load checkpoint comparison.",e&&(e.textContent="");return}if(A("bar-chart",t.lastUpdated),k("bar-chart",t.source),I(n,t.groups,{title:`${t.metric} (${t.unit})`,description:(r=t.summary)!=null?r:`${t.metric} comparison`}),e){let a=t.groups.filter(o=>o.context).map(o=>`${o.label}: ${o.context}`).join(" ");e.innerHTML=`<strong>${t.metric}:</strong> ${(s=t.summary)!=null?s:a}`}}async function O(){let n=await w("data/global-stage.json");n&&(A("global-stage",n.lastUpdated),k("global-stage",n.source))}function q(){let n=document.querySelectorAll("[data-chart]");if(n.length===0)return;let e={sparkline:D,"bar-chart":j};if(!("IntersectionObserver"in window)){n.forEach(r=>{var o;let s=(o=r.dataset.chart)!=null?o:"",a=e[s];a&&a()});return}let t=new IntersectionObserver((r,s)=>{r.forEach(a=>{var u;if(!a.isIntersecting)return;let o=a.target,l=(u=o.dataset.chart)!=null?u:"",i=e[l];i&&(i(),s.unobserve(o))})},{threshold:.2});n.forEach(r=>t.observe(r))}window.addEventListener("DOMContentLoaded",()=>{P(),O(),q()});
