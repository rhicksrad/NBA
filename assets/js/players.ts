
type PlayersIndex = {
  fetched_at: string;
  source: string;
  count: number;
  players: Array<{
    id: number;
    name: string;
    team_abbr: string;
    position: string | null;
    jersey: string | null;
    height: string | null;
    weight: string | null;
  }>;
};

type PlayerRow = PlayersIndex["players"][number];

type AppState = {
  index: PlayersIndex | null;
  loading: boolean;
  error: string | null;
  searchTerm: string;
  teamFilter: string;
  anchorApplied: boolean;
};

async function loadIndex(): Promise<PlayersIndex> {
  const response = await fetch(`data/players_index.json?cb=${Date.now()}`);
  if (!response.ok) {
    throw new Error(`Request failed with status ${response.status}`);
  }
  return (await response.json()) as PlayersIndex;
}

const app = document.getElementById("players-app");

if (!app) {
  throw new Error("Missing #players-app container");
}

const params = new URLSearchParams(window.location.search);
const initialTeam = (params.get("team") ?? "").toUpperCase();
const initialSearch = params.get("search") ?? "";

const state: AppState = {
  index: null,
  loading: true,
  error: null,
  searchTerm: initialSearch,
  teamFilter: initialTeam,
  anchorApplied: false,
};

function escapeHtml(value: string): string {
  return value.replace(/[&<>"']/g, (char) => {
    switch (char) {
      case "&":
        return "&amp;";
      case "<":
        return "&lt;";
      case ">":
        return "&gt;";
      case '"':
        return "&quot;";
      case "'":
        return "&#39;";
      default:
        return char;
    }
  });
}

function formatRelativeTime(iso: string): string {
  const timestamp = Date.parse(iso);
  if (Number.isNaN(timestamp)) {
    return iso;
  }

  const now = Date.now();
  const diffSeconds = Math.round((timestamp - now) / 1000);
  const divisions: Array<{ amount: number; unit: Intl.RelativeTimeFormatUnit }> = [
    { amount: 60, unit: "second" },
    { amount: 60, unit: "minute" },
    { amount: 24, unit: "hour" },
    { amount: 7, unit: "day" },
    { amount: 4.34524, unit: "week" },
    { amount: 12, unit: "month" },
    { amount: Number.POSITIVE_INFINITY, unit: "year" },
  ];

  const formatter = new Intl.RelativeTimeFormat("en", { numeric: "auto" });
  let duration = diffSeconds;
  for (const division of divisions) {
    if (Math.abs(duration) < division.amount) {
      return formatter.format(Math.round(duration), division.unit);
    }
    duration /= division.amount;
  }
  return formatter.format(Math.round(duration), "year");
}

function matchesSearch(player: PlayerRow, query: string): boolean {
  if (!query) {
    return true;
  }
  const lower = query.toLowerCase();
  const name = player.name.toLowerCase();
  const jersey = player.jersey ? `#${player.jersey}`.toLowerCase() : "";
  return name.includes(lower) || jersey.includes(lower);
}

function updateUrl(paramsToSet: Record<string, string | null>) {
  const url = new URL(window.location.href);
  Object.entries(paramsToSet).forEach(([key, value]) => {
    if (value && value.length) {
      url.searchParams.set(key, value);
    } else {
      url.searchParams.delete(key);
    }
  });
  window.history.replaceState({}, "", url.toString());
}

function teamLabel(abbr: string): string {
  if (abbr === "FA") {
    return "Free agents";
  }
  return abbr;
}

function sortTeams(a: string, b: string): number {
  if (a === "FA" && b !== "FA") {
    return 1;
  }
  if (b === "FA" && a !== "FA") {
    return -1;
  }
  return a.localeCompare(b);
}

function buildTeams(players: PlayerRow[]): Array<{ abbr: string; players: PlayerRow[] }> {
  const map = new Map<string, PlayerRow[]>();
  for (const player of players) {
    const abbr = player.team_abbr || "FA";
    const roster = map.get(abbr);
    if (roster) {
      roster.push(player);
    } else {
      map.set(abbr, [player]);
    }
  }
  return [...map.entries()]
    .map(([abbr, roster]) => ({ abbr, players: roster }))
    .sort((lhs, rhs) => sortTeams(lhs.abbr, rhs.abbr));
}

function renderLoading() {
  app.innerHTML = `
    <div class="roster-status">
      <p>Loading active rosters…</p>
    </div>
  `;
}

function renderError(message: string) {
  app.innerHTML = `
    <div class="roster-status roster-status--error">
      <p>${escapeHtml(message)}</p>
      <button type="button" class="roster-button" data-roster-retry>Retry</button>
    </div>
  `;
  const retry = app.querySelector<HTMLButtonElement>("[data-roster-retry]");
  if (retry) {
    retry.addEventListener("click", () => fetchIndex());
  }
}

function renderDoc(index: PlayersIndex) {
  const teams = buildTeams(index.players);
  const selectedTeam = state.teamFilter;
  if (selectedTeam && !teams.some((team) => team.abbr === selectedTeam)) {
    state.teamFilter = "";
    updateUrl({ team: null });
  }
  const effectiveTeam = state.teamFilter;
  const visibleTeams = teams.filter(
    (team) => !effectiveTeam || team.abbr === effectiveTeam,
  );
  const teamOptions = ["", ...teams.map((team) => team.abbr)];

  const hasTeams = teams.length > 0;
  const lastUpdatedText = hasTeams
    ? formatRelativeTime(index.fetched_at)
    : "not yet available";
  const timestampTitle = hasTeams
    ? new Date(index.fetched_at).toLocaleString()
    : "No roster snapshot cached yet";
  const sourceLabel = index.source?.trim() || "Unknown";

  const headerHtml = `
    <div class="roster-controls">
      <div class="roster-controls__filters">
        <label class="roster-controls__field">
          <span class="roster-controls__label">Search</span>
          <input
            id="roster-search"
            class="roster-input"
            type="search"
            placeholder="Search by name or jersey"
            value="${escapeHtml(state.searchTerm)}"
            autocomplete="off"
          />
        </label>
        <label class="roster-controls__field">
          <span class="roster-controls__label">Team</span>
          <select id="roster-team" class="roster-select">
            ${teamOptions
              .map((code) => {
                const label = code || "All teams";
                const selected = code === effectiveTeam ? "selected" : "";
                return `<option value="${code}">${label}</option>`;
              })
              .join("")}
          </select>
        </label>
      </div>
      <div class="roster-controls__meta">
        <small title="${timestampTitle}">
          Last updated: ${lastUpdatedText} • Source: ${escapeHtml(sourceLabel)}
        </small>
        <button type="button" class="roster-button" data-roster-refresh>Refresh</button>
      </div>
    </div>
  `;

  const sections = visibleTeams
    .map((team) => {
      const players = team.players.filter((player) =>
        matchesSearch(player, state.searchTerm),
      );
      const items = players
        .map((player) => {
          const jersey = player.jersey ? `#${player.jersey}` : "";
          const pieces = [player.position ?? "", jersey].filter(Boolean).join(" · ");
          const meta = [player.height ?? "", player.weight ? `${player.weight} lb` : ""]
            .filter(Boolean)
            .join(" • ");
          return `
            <li class="roster-player">
              <span class="roster-player__name">${escapeHtml(player.name)}</span>
              ${pieces ? `<span class="roster-player__role">${escapeHtml(pieces)}</span>` : ""}
              ${meta ? `<span class="roster-player__meta">${escapeHtml(meta)}</span>` : ""}
            </li>
          `;
        })
        .join("");

      const emptyMessage = players.length
        ? ""
        : `<li class="roster-player roster-player--empty">No players match this filter.</li>`;

      const subtitle = `${escapeHtml(teamLabel(team.abbr))} · ${players.length} players`;

      return `
        <section class="roster-team" data-team-anchor="${team.abbr}">
          <header class="roster-team__header">
            <h3 id="team-${team.abbr}">${team.abbr}</h3>
            <p>${subtitle}</p>
          </header>
          <ul class="roster-list">
            ${items || emptyMessage}
          </ul>
        </section>
      `;
    })
    .join("");

  let noTeamsMessage = "";
  if (!hasTeams) {
    noTeamsMessage = `<div class="roster-status roster-status--empty"><p>Rosters are not cached yet. Use Refresh to try again.</p></div>`;
  } else if (!visibleTeams.length) {
    noTeamsMessage = `<div class="roster-status roster-status--empty"><p>No teams match the current filter.</p></div>`;
  }

  app.innerHTML = `${headerHtml}<div class="roster-teams">${sections}${noTeamsMessage}</div>`;

  const searchInput = document.getElementById("roster-search") as HTMLInputElement | null;
  const teamSelect = document.getElementById("roster-team") as HTMLSelectElement | null;
  const refreshButton = app.querySelector<HTMLButtonElement>("[data-roster-refresh]");

  if (searchInput) {
    searchInput.addEventListener("input", (event) => {
      const value = (event.target as HTMLInputElement).value;
      state.searchTerm = value;
      updateUrl({ search: value });
      render();
    });
  }

  if (teamSelect) {
    teamSelect.addEventListener("change", (event) => {
      const value = (event.target as HTMLSelectElement).value.toUpperCase();
      state.teamFilter = value;
      state.anchorApplied = !value;
      updateUrl({ team: value });
      render();
    });
  }

  if (refreshButton) {
    refreshButton.addEventListener("click", () => fetchIndex());
  }

  if (!state.anchorApplied && state.teamFilter) {
    const anchor = app.querySelector(`[data-team-anchor="${state.teamFilter}"]`);
    if (anchor) {
      anchor.scrollIntoView({ behavior: "smooth", block: "start" });
      state.anchorApplied = true;
    }
  }
}

function render() {
  if (state.loading) {
    renderLoading();
    return;
  }
  if (state.error) {
    renderError(state.error);
    return;
  }
  if (state.index) {
    renderDoc(state.index);
  }
}

async function fetchIndex() {
  state.loading = true;
  state.error = null;
  render();
  try {
    const index = await loadIndex();
    if (!index || !Array.isArray(index.players)) {
      throw new Error("Malformed players index payload");
    }
    state.index = index;
    state.loading = false;
    render();
  } catch (error) {
    state.loading = false;
    state.error = error instanceof Error ? error.message : "Unable to load players.";
    render();
  }
}

fetchIndex();
