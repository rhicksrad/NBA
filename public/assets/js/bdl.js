export const API = 'https://bdlproxy.hicksrch.workers.dev/bdl';

function makeLimiter({ maxConcurrent = 1, minIntervalMs = 300 } = {}) {
  let active = 0;
  const queue = [];
  let lastStart = 0;

  const scheduleNext = () => {
    if (!queue.length || active >= maxConcurrent) return;
    const task = queue.shift();
    if (!task) return;
    run(task.fn, task.resolve, task.reject);
  };

  const run = (fn, resolve, reject) => {
    const now = Date.now();
    const wait = Math.max(0, minIntervalMs - (now - lastStart));
    lastStart = now + wait;
    setTimeout(async () => {
      active += 1;
      try {
        const value = await fn();
        resolve(value);
      } catch (error) {
        reject(error);
      } finally {
        active -= 1;
        scheduleNext();
      }
    }, wait);
  };

  return (fn) =>
    new Promise((resolve, reject) => {
      if (active < maxConcurrent) {
        run(fn, resolve, reject);
      } else {
        queue.push({ fn, resolve, reject });
      }
    });
}

const limit = makeLimiter({ maxConcurrent: 1, minIntervalMs: 300 });
const memo = new Map();

function shouldMemoize(init = {}) {
  const method = typeof init.method === 'string' ? init.method.toUpperCase() : 'GET';
  if (method !== 'GET') return false;
  if (init.cache && init.cache.toLowerCase() === 'no-store') return false;
  return true;
}

function parseRetryAfter(value) {
  if (!value) return null;
  const numeric = Number(value);
  if (Number.isFinite(numeric)) {
    return Math.max(0, numeric * 1000);
  }
  const parsed = Date.parse(value);
  if (!Number.isNaN(parsed)) {
    return Math.max(0, parsed - Date.now());
  }
  return null;
}

async function fetchJsonWithRetry(url, init = {}) {
  const headers = new Headers({ Accept: 'application/json' });
  if (init.headers instanceof Headers) {
    init.headers.forEach((value, key) => headers.set(key, value));
  } else if (init.headers && typeof init.headers === 'object') {
    for (const [key, value] of Object.entries(init.headers)) {
      if (value != null) headers.set(key, value);
    }
  }

  const baseInit = {
    ...init,
    headers,
    method: init.method ?? 'GET'
  };

  for (let attempt = 1; attempt <= 6; attempt += 1) {
    const response = await fetch(url, baseInit);
    if (response.status === 429) {
      const retryAfter = parseRetryAfter(response.headers.get('Retry-After'));
      const backoff = retryAfter != null ? retryAfter : Math.min(3000 * attempt, 15000);
      const jitter = Math.floor(Math.random() * 400);
      await new Promise((resolve) => setTimeout(resolve, backoff + jitter));
      continue;
    }
    if (!response.ok) {
      const text = await response.text().catch(() => '');
      const path = (() => {
        try {
          return new URL(url).pathname;
        } catch {
          return url;
        }
      })();
      throw new Error(`BDL ${response.status} for ${path}${text ? `: ${text.slice(0, 120)}` : ''}`);
    }
    return response.json();
  }
  throw new Error('BDL proxy retried too many times after 429 responses.');
}

export async function bdl(path, init = {}) {
  const url = `${API}${path}`;
  const memoKey = shouldMemoize(init) ? url : null;
  if (memoKey && memo.has(memoKey)) {
    return memo.get(memoKey);
  }

  const task = limit(() => fetchJsonWithRetry(url, init));
  if (!memoKey) {
    return task;
  }

  memo.set(memoKey, task);
  try {
    const result = await task;
    memo.set(memoKey, result);
    return result;
  } catch (error) {
    memo.delete(memoKey);
    throw error;
  }
}

export async function fetchSeasonAggregate({ season, playerId, postseason = false, signal } = {}) {
  const params = new URLSearchParams({
    season: String(season),
    player_id: String(playerId)
  });
  if (postseason) params.set('postseason', 'true');
  return bdl(`/v1/season_averages?${params.toString()}`, { signal });
}
