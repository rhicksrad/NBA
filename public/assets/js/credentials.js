const MAX_ATTEMPTS = 5;
const POLL_DELAY_MS = 10;
let readinessPromise = null;

function normalizeAuthorization(key) {
  const trimmed = String(key ?? '').trim();
  if (!trimmed) {
    return null;
  }
  if (/^Bearer\s+/i.test(trimmed)) {
    return trimmed;
  }
  return `Bearer ${trimmed}`;
}

export function resolveBdlKeySync() {
  if (typeof document !== 'undefined') {
    const meta = document.querySelector('meta[name="bdl-api-key"]');
    const content = meta?.getAttribute('content')?.trim();
    if (content) {
      return content;
    }
  }

  const inline = globalThis?.BDL_CREDENTIALS?.key;
  if (typeof inline === 'string' && inline.trim()) {
    return inline.trim();
  }

  return null;
}

export async function ensureBdlKeyReady() {
  if (!readinessPromise) {
    readinessPromise = (async () => {
      for (let attempt = 0; attempt < MAX_ATTEMPTS; attempt += 1) {
        const key = resolveBdlKeySync();
        if (key) {
          return key;
        }
        await new Promise((resolve) => setTimeout(resolve, POLL_DELAY_MS));
      }
      throw new Error("Ball Don't Lie API key missing");
    })();
  }

  return readinessPromise;
}

export async function authHeaders() {
  const key = await ensureBdlKeyReady();
  const authorization = normalizeAuthorization(key);
  if (!authorization) {
    throw new Error("Ball Don't Lie API key missing");
  }
  return { Authorization: authorization };
}
