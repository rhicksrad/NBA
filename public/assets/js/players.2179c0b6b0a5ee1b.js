async function H(){let e=await fetch(`data/players_index.json?cb=${Date.now()}`);if(!e.ok)throw new Error(`Request failed with status ${e.status}`);return await e.json()}var u=document.getElementById("players-app");if(!u)throw new Error("Missing #players-app container");var S=new URLSearchParams(window.location.search),E,B=((E=S.get("team"))!=null?E:"").toUpperCase(),A,N=(A=S.get("search"))!=null?A:"",r={index:null,loading:!0,error:null,searchTerm:N,teamFilter:B,anchorApplied:!1};function m(e){return e.replace(/[&<>"']/g,t=>{switch(t){case"&":return"&amp;";case"<":return"&lt;";case">":return"&gt;";case'"':return"&quot;";case"'":return"&#39;";default:return t}})}function U(e){let t=Date.parse(e);if(Number.isNaN(t))return e;let n=Date.now(),a=Math.round((t-n)/1e3),l=[{amount:60,unit:"second"},{amount:60,unit:"minute"},{amount:24,unit:"hour"},{amount:7,unit:"day"},{amount:4.34524,unit:"week"},{amount:12,unit:"month"},{amount:Number.POSITIVE_INFINITY,unit:"year"}],h=new Intl.RelativeTimeFormat("en",{numeric:"auto"}),c=a;for(let d of l){if(Math.abs(c)<d.amount)return h.format(Math.round(c),d.unit);c/=d.amount}return h.format(Math.round(c),"year")}function C(e,t){if(!t)return!0;let n=t.toLowerCase(),a=e.name.toLowerCase(),l=e.jersey?`#${e.jersey}`.toLowerCase():"";return a.includes(n)||l.includes(n)}function y(e){let t=new URL(window.location.href);Object.entries(e).forEach(([n,a])=>{a&&a.length?t.searchParams.set(n,a):t.searchParams.delete(n)}),window.history.replaceState({},"",t.toString())}function D(e){return e==="FA"?"Free agents":e}function k(e,t){return e==="FA"&&t!=="FA"?1:t==="FA"&&e!=="FA"?-1:e.localeCompare(t)}function q(e){let t=new Map;for(let n of e){let a=n.team_abbr||"FA",l=t.get(a);l?l.push(n):t.set(a,[n])}return[...t.entries()].map(([n,a])=>({abbr:n,players:a})).sort((n,a)=>k(n.abbr,a.abbr))}function O(){u.innerHTML=`
    <div class="roster-status">
      <p>Loading active rosters\u2026</p>
    </div>
  `}function V(e){u.innerHTML=`
    <div class="roster-status roster-status--error">
      <p>${m(e)}</p>
      <button type="button" class="roster-button" data-roster-retry>Retry</button>
    </div>
  `;let t=u.querySelector("[data-roster-retry]");t&&t.addEventListener("click",()=>g())}function Y(e){let t=q(e.players),n=r.teamFilter;n&&!t.some(s=>s.abbr===n)&&(r.teamFilter="",y({team:null}));let a=r.teamFilter,l=t.filter(s=>!a||s.abbr===a),h=["",...t.map(s=>s.abbr)],c=t.length>0,d=c?U(e.fetched_at):"not yet available",F=c?new Date(e.fetched_at).toLocaleString():"No roster snapshot cached yet",M=`
    <div class="roster-controls">
      <div class="roster-controls__filters">
        <label class="roster-controls__field">
          <span class="roster-controls__label">Search</span>
          <input
            id="roster-search"
            class="roster-input"
            type="search"
            placeholder="Search by name or jersey"
            value="${m(r.searchTerm)}"
            autocomplete="off"
          />
        </label>
        <label class="roster-controls__field">
          <span class="roster-controls__label">Team</span>
          <select id="roster-team" class="roster-select">
            ${h.map(s=>{let o=s||"All teams",T=s===a?"selected":"";return`<option value="${s}">${o}</option>`}).join("")}
          </select>
        </label>
      </div>
      <div class="roster-controls__meta">
        <small title="${F}">
          Last updated: ${d} \u2022 Source: BallDontLie
        </small>
        <button type="button" class="roster-button" data-roster-refresh>Refresh</button>
      </div>
    </div>
  `,j=l.map(s=>{let o=s.players.filter(i=>C(i,r.searchTerm)),T=o.map(i=>{var L,I;let x=i.jersey?`#${i.jersey}`:"",$=[(L=i.position)!=null?L:"",x].filter(Boolean).join(" \xB7 "),_=[(I=i.height)!=null?I:"",i.weight?`${i.weight} lb`:""].filter(Boolean).join(" \u2022 ");return`
            <li class="roster-player">
              <span class="roster-player__name">${m(i.name)}</span>
              ${$?`<span class="roster-player__role">${m($)}</span>`:""}
              ${_?`<span class="roster-player__meta">${m(_)}</span>`:""}
            </li>
          `}).join(""),P=o.length?"":'<li class="roster-player roster-player--empty">No players match this filter.</li>',R=`${m(D(s.abbr))} \xB7 ${o.length} players`;return`
        <section class="roster-team" data-team-anchor="${s.abbr}">
          <header class="roster-team__header">
            <h3 id="team-${s.abbr}">${s.abbr}</h3>
            <p>${R}</p>
          </header>
          <ul class="roster-list">
            ${T||P}
          </ul>
        </section>
      `}).join(""),f="";c?l.length||(f='<div class="roster-status roster-status--empty"><p>No teams match the current filter.</p></div>'):f='<div class="roster-status roster-status--empty"><p>Rosters are not cached yet. Use Refresh to try again.</p></div>',u.innerHTML=`${M}<div class="roster-teams">${j}${f}</div>`;let b=document.getElementById("roster-search"),w=document.getElementById("roster-team"),v=u.querySelector("[data-roster-refresh]");if(b&&b.addEventListener("input",s=>{let o=s.target.value;r.searchTerm=o,y({search:o}),p()}),w&&w.addEventListener("change",s=>{let o=s.target.value.toUpperCase();r.teamFilter=o,r.anchorApplied=!o,y({team:o}),p()}),v&&v.addEventListener("click",()=>g()),!r.anchorApplied&&r.teamFilter){let s=u.querySelector(`[data-team-anchor="${r.teamFilter}"]`);s&&(s.scrollIntoView({behavior:"smooth",block:"start"}),r.anchorApplied=!0)}}function p(){if(r.loading){O();return}if(r.error){V(r.error);return}r.index&&Y(r.index)}async function g(){r.loading=!0,r.error=null,p();try{let e=await H();if(!e||!Array.isArray(e.players))throw new Error("Malformed players index payload");r.index=e,r.loading=!1,p()}catch(e){r.loading=!1,r.error=e instanceof Error?e.message:"Unable to load players.",p()}}g();
