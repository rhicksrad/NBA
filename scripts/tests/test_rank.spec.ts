import { describe, expect, it } from "vitest";
import { rankTeams } from "../generate/rank.js";
import { RankedTeam, TeamRecord, PlayerRecord, InjuryRecord } from "../lib/types.js";

describe("rankTeams", () => {
  it("produces a deterministic ranking snapshot", () => {
    const teams: TeamRecord[] = [
      {
        teamId: "1",
        tricode: "AAA",
        market: "Alpha",
        name: "Anchors",
        coach: { name: "Casey" },
        lastSeasonWins: 50,
        lastSeasonSRS: 4.2,
        roster: [
          { name: "Alex Prime", position: "G", teamId: "1", teamTricode: "AAA" },
          { name: "Brian Bolt", position: "F", teamId: "1", teamTricode: "AAA" },
        ],
        keyAdditions: ["Alex Prime"],
        keyLosses: [],
        notes: [],
      },
      {
        teamId: "2",
        tricode: "BBB",
        market: "Bravo",
        name: "Blazers",
        coach: { name: "Dana", isNew: true },
        lastSeasonWins: 38,
        lastSeasonSRS: -0.5,
        roster: [
          { name: "Chris Craft", position: "G", teamId: "2", teamTricode: "BBB" },
          { name: "Devon Dash", position: "F", teamId: "2", teamTricode: "BBB" },
        ],
        keyAdditions: ["Chris Craft"],
        keyLosses: ["Jamie Jump"],
        notes: [],
      },
      {
        teamId: "3",
        tricode: "CCC",
        market: "Charlie",
        name: "Chargers",
        coach: { name: "Evan" },
        lastSeasonWins: 28,
        lastSeasonSRS: -3.1,
        roster: [
          { name: "Frank Flash", position: "G", teamId: "3", teamTricode: "CCC" },
        ],
        keyAdditions: [],
        keyLosses: ["Kelly Kinetics"],
        notes: [],
      },
    ];

    const players: PlayerRecord[] = teams.flatMap((team) => team.roster);
    const injuries: InjuryRecord[] = [
      { playerName: "Chris Craft", status: "Day-to-day", severity: "medium" },
    ];

    const rankings: RankedTeam[] = rankTeams({ teams, players, injuries });

    expect(rankings).toMatchSnapshot();
  });
});
