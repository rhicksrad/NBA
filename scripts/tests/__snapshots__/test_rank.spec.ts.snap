// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rankTeams > produces a deterministic ranking snapshot 1`] = `
[
  {
    "rank": 1,
    "score": 0.733808,
    "statusLine": "Uptrend | Staff intact | Added ball handling | Health risk: low",
    "tricode": "AAA",
  },
  {
    "rank": 2,
    "score": 0.581442,
    "statusLine": "Rebuild | Staff intact | Banking on internal growth | Health risk: low",
    "tricode": "CCC",
  },
  {
    "rank": 3,
    "score": 0.567404,
    "statusLine": "Holding | New coach | Added ball handling | Health risk: medium",
    "tricode": "BBB",
  },
]
`;
